src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -1275,6 +1275,20 @@ private async ValueTask SendDataAsync(ReadOnlyMemory<byte> buffer, CancellationT
                         await _connection.SendStreamDataAsync(StreamId, current, flush, _requestBodyCancellationSource.Token).ConfigureAwait(false);
                     }
                 }
+                catch (OperationCanceledException e) when (e.CancellationToken == _requestBodyCancellationSource.Token)
+                {
+                    if (_resetException is Exception resetException)
+                    {
+                        if (_canRetry)
```

Comment:
I think we need to take the lock around checking _resetException and _canRetry. Otherwise we may see an inconsistent view of these two values.

Reply:
Good point. I overlooked it. Fixed.